src/arrow1/reader_async.rs

```diff
@@ -1,15 +1,670 @@
+                    let fetch_bytes = &fetched[idx];
+
+                    let start = range.start - fetch_range.start;
+                    let end = range.end - fetch_range.start;
+                    fetch_bytes.slice(start..end)
+                })
+                .collect())
+        }
+        .boxed()
+    }
+
+    fn get_metadata(&mut self) -> BoxFuture<'_, parquet::errors::Result<Arc<ParquetMetaData>>> {
+        async move {
+            let meta = fetch_parquet_metadata(self.url.as_str(), &self.client, None).await?;
+            Ok(Arc::new(meta))
+        }
+        .boxed()
+    }
+}
+
+#[wasm_bindgen]
+pub struct AsyncParquetLocalFile {
+    reader: JsFileReader,
+    meta: ArrowReaderMetadata,
+    batch_size: usize,
+    projection_mask: Option<ProjectionMask>,
+}
+
+impl SharedIO<JsFileReader> for AsyncParquetLocalFile {}
+
+#[wasm_bindgen]
+impl AsyncParquetLocalFile {
+    #[wasm_bindgen(constructor)]
+    pub async fn new(handle: web_sys::File) -> WasmResult<AsyncParquetLocalFile> {
+        let mut reader = JsFileReader::new(handle, 1024);
+        let meta = ArrowReaderMetadata::load_async(&mut reader, Default::default()).await?;
+        Ok(Self {
+            reader,
+            meta,
+            batch_size: 1024,
+            projection_mask: None,
+        })
+    }
+
+    #[wasm_bindgen(js_name = withBatchSize)]
+    pub fn with_batch_size(self, batch_size: usize) -> Self {
+        Self { batch_size, ..self }
+    }
+
+    #[wasm_bindgen(js_name = selectColumns)]
+    pub fn select_columns(self, columns: Vec<String>) -> WasmResult<AsyncParquetLocalFile> {
+        let pq_schema = self.meta.parquet_schema();
+        let projection_mask = Some(generate_projection_mask(columns, pq_schema)?);
+        Ok(Self {
+            projection_mask,
+            ..self
+        })
+    }
+
+    #[wasm_bindgen]
+    pub fn metadata(&self) -> WasmResult<crate::arrow1::metadata::ParquetMetaData> {
+        Ok(self.meta.metadata().as_ref().to_owned().into())
+    }
+
+    #[wasm_bindgen(js_name = readRowGroup)]
+    pub async fn read_row_group(&self, i: usize) -> WasmResult<Table> {
+        let inner = self
+            .inner_read_row_group(
+                &self.reader,
+                &self.meta,
+                &self.batch_size,
+                &self.projection_mask,
+                i,
+            )
+            .await
+            .unwrap();
+        Ok(inner)
+    }
+
+    #[wasm_bindgen]
+    pub async fn stream(
+        &self,
+        concurrency: Option<usize>,
+    ) -> WasmResult<wasm_streams::readable::sys::ReadableStream> {
+        self.inner_stream(
+            concurrency,
+            &self.meta,
+            &self.reader,
+            &self.batch_size,
+            &self.projection_mask,
+        )
+        .await
+    }
+}
+
+#[derive(Debug, Clone)]
+struct WrappedFile {
+    inner: Arc<web_sys::File>,
+}
+
+unsafe impl Send for WrappedFile {}
+unsafe impl Sync for WrappedFile {}
```

Comment:
can we add some reasoning here, or a `Safety` comment? I didn't think any JS objects would be safe to transfer across threads without `SharedArrayBuffer`, but also it's confusing because we're not working across threads here.

I thought the messaging of `oneshot::channel` didn't need `Send` and `Sync`?

Reply:
We need this because AsyncParquetFile has a `Sync` bound?

Reply:
The AsyncFileReader trait, yeah - I've narrowed down the unsafe part to just the `unsafe impl Send` part (pushing shortly).

The difference between AsyncParquetFile and AsyncParquetLocalFile is the hold on the non-Send value (I've tried quite a few variations, `Arc<Mutex<File>>` included), the AsyncFileReader just outright bans non-Send readers (even if the async runtime is explicitly single-thread or non-workstealing).

Apparently it's possible to feature flag trait constraints, so situations like this (async traits on anything that holds a JSValue reference) could be solved by opting into a 'non_threadsafe' feature.

I'll add a safety note to both the unsafe block and the AsyncParquetLocalFile along the lines of "Don't use in multi-threaded environments".